client/src/actions/annotation.ts

```diff
@@ -0,0 +1,643 @@
+import type { Genesets } from "../reducers/genesets";
+import {
+  AnnotationColumnSchema,
+  Category,
+  Field,
+} from "../common/types/schema";
+import { AnyArray } from "../common/types/arraytypes";
+
+type ThunkResult<R = void> = ThunkAction<R, RootState, never, AnyAction>;
+
+type ColumnValueCtor = new (length: number) => AnyArray;
+
+const { isUserAnnotation } = AnnotationsHelpers;
+
+export const annotationCreateCategoryAction =
+  (newCategoryName: string, categoryToDuplicate?: string | null): ThunkResult =>
+  async (dispatch: AppDispatch, getState: GetState) => {
+    const state = getState();
+    const { annoMatrix: prevAnnoMatrix, obsCrossfilter: prevObsCrossfilter } =
+      state;
+    if (!prevAnnoMatrix || !prevObsCrossfilter) return;
+
+    const trimmedName = newCategoryName.trim();
+    if (!trimmedName) return;
+
+    const { schema } = prevAnnoMatrix;
+    if (schema.annotations.obsByName[trimmedName]) {
+      throw new Error("name collision on annotation category create");
+    }
+
+    let initialValue: AnyArray | Category;
+    let ctor: ColumnValueCtor;
+    let newSchema: AnnotationColumnSchema;
+
+    if (categoryToDuplicate) {
+      const sourceSchema = schema.annotations.obsByName[categoryToDuplicate];
+      if (!sourceSchema) {
+        throw new Error("categoryToDuplicate does not exist");
+      }
+      if (
+        sourceSchema.type !== "string" &&
+        sourceSchema.type !== "categorical"
+      ) {
+        throw new Error(
+          "categoryToDuplicate does not exist or has invalid type"
+        );
+      }
+
+      const catDf = await prevAnnoMatrix
+        .base()
+        .fetch(Field.obs, categoryToDuplicate);
+      const column = catDf.col(categoryToDuplicate);
+      const columnArray = column.asArray() as AnyArray;
+      const categories = column.summarizeCategorical().categories.slice();
+      if (!categories.includes(globals.unassignedCategoryLabel)) {
+        categories.push(globals.unassignedCategoryLabel);
+      }
+      initialValue = columnArray;
+      ctor = (columnArray.constructor as ColumnValueCtor) ?? Array;
+      newSchema = {
+        ...sourceSchema,
+        name: trimmedName,
+        categories,
+        writable: true,
+      };
+    } else {
+      initialValue = globals.unassignedCategoryLabel;
+      ctor = Array as unknown as ColumnValueCtor;
+      newSchema = {
+        name: trimmedName,
+        type: "categorical",
+        categories: [globals.unassignedCategoryLabel],
+        writable: true,
+      };
+    }
+
+    const obsCrossfilter = prevObsCrossfilter.addObsColumn(
+      newSchema,
+      ctor,
+      initialValue
+    );
+
+    dispatch({
+      type: "annotation: create category",
+      data: trimmedName,
+      categoryToDuplicate: categoryToDuplicate ?? null,
+      annoMatrix: obsCrossfilter.annoMatrix,
+      obsCrossfilter,
+    });
+  };
+
+export const annotationRenameCategoryAction =
```

Comment:
Rename action optimistically updates state then gates the backend call on `config.parameters.annotations`; calling that out so folks know the autosave reset happens regardless of network outcome.